libnmstate/nm/nmclient.py

```diff
@@ -220,3 +247,67 @@ def _register_first_action(self):
     def _execute_action_once(self, _):
         self.execute_next_action()
         return False
+
+
+class _NmClient:
+    def __init__(self):
+        self._client = NM.Client.new(None)
+        logging.debug("NM.Client created")
+
+    @property
+    def client(self):
+        return self._client
+
+    def delete_client(self):
+        if not self._client:
+            return
+        try:
+            ref_count = sys.getrefcount(self._client)
+            if ref_count != 2:
+                # The getrefcount() hold 1 reference and self.client itself
+                # hold 1 reference.
+                raise error.NmstateInternalError(
+                    "_NmClient.delete_client() cannot release resources "
+                    "of NM.Client duo to other unreleased use."
+                )
+            self._nmclient_cleanup()
+        except Exception:
+            pass
+
+    def __del__(self):
+        if self._client:
+            raise error.NmstateInternalError(
+                "_NmClient.delete_client() should be explicitly called "
+                "before python GC clean up the object."
+            )
+
+    def _nmclient_cleanup(self):
+        if self._client:
+            is_done = []
+            try:
+                context = self.client.get_main_context()
+                self.client.get_context_busy_watcher().weak_ref(
+                    lambda: is_done.append(1)
+                )
+            except AttributeError:
+                context = GLib.MainContext.default()
+                is_done.append(1)
+
+            self._client = None
+
+            while context.iteration(False):
+                pass
+
+            if not is_done:
+                logging.debug(
+                    "context.iteration() does not delete "
+                    "the context_busy_watcher, "
+                    "waiting 50 milliseconds"
+                )
+                timeout_source = GLib.timeout_source_new(50)
+                timeout_source.set_callback(lambda x: is_done.append(1))
+                timeout_source.attach(context)
+                while not is_done:
+                    context.iteration(True)
+                timeout_source.destroy()
```

Comment:
The timeout_source should probably be treated as a resource with a contextmanager, otherwise in case of a failure in the meddle it will not cleanup.

Reply:
Wrappered with `try...finally`.